xbmc/platform/win32/IMMNotificationClient.h

```diff
@@ -99,68 +93,91 @@ class CMMNotificationClient : public IMMNotificationClient
       break;
     case eCommunications:
       pszRole = "eCommunications";
-      NotifyAE();
       break;
     }
 
-    CLog::Log(LOGDEBUG, "{}: New default device: flow = {}, role = {}", __FUNCTION__, pszFlow,
-              pszRole);
+    CLog::LogF(LOGDEBUG, "New default device: flow = {}, role = {}", pszFlow, pszRole);
+
     return S_OK;
   }
 
   HRESULT STDMETHODCALLTYPE OnDeviceAdded(LPCWSTR pwstrDeviceId)
   {
-    CLog::Log(LOGDEBUG, "{}: Added device: {}", __FUNCTION__, FromW(pwstrDeviceId));
+    CLog::LogF(LOGDEBUG, "Added device: {}", FromW(pwstrDeviceId));
     NotifyAE();
     return S_OK;
   }
 
   HRESULT STDMETHODCALLTYPE OnDeviceRemoved(LPCWSTR pwstrDeviceId)
   {
-    CLog::Log(LOGDEBUG, "{}: Removed device: {}", __FUNCTION__, FromW(pwstrDeviceId));
+    CLog::LogF(LOGDEBUG, "Removed device: {}", FromW(pwstrDeviceId));
     NotifyAE();
     return S_OK;
   }
 
   HRESULT STDMETHODCALLTYPE OnDeviceStateChanged(LPCWSTR pwstrDeviceId, DWORD dwNewState)
   {
-    std::string pszState = "?????";
+    std::string pszState;
+    bool deviceIsDown{false};
 
     switch (dwNewState)
     {
     case DEVICE_STATE_ACTIVE:
       pszState = "ACTIVE";
+      deviceIsDown = false;
       break;
     case DEVICE_STATE_DISABLED:
       pszState = "DISABLED";
+      deviceIsDown = true;
       break;
     case DEVICE_STATE_NOTPRESENT:
       pszState = "NOTPRESENT";
+      deviceIsDown = true;
       break;
     case DEVICE_STATE_UNPLUGGED:
       pszState = "UNPLUGGED";
+      deviceIsDown = true;
       break;
     }
-    CLog::Log(LOGDEBUG, "{}: New device state is DEVICE_STATE_{}", __FUNCTION__, pszState);
-    NotifyAE();
+
+    CLog::LogF(LOGDEBUG, "New device state is DEVICE_STATE_{}", pszState);
+
+    if (deviceIsDown != m_deviceIsDown)
```

Comment:
I'm not sure that this will work. In common case you compare state of different devices. 
1. device1 became active (not used by AE) -> m_deviceIsDown = false; -> NotifyAE();
2. device2 becomes active (used by AE) -> deviceIsDown == m_deviceIsDown -> nothing happened.

or I skip somethig important?